### PR TITLE
chapter_12

### PR DESCRIPTION
**Pull request by chuan97sun**

### CODE REVIEW

**Comment by chuan97sun** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: %i(edit update)
+  before_action :valid_user, only: %i(edit update)
+  before_action :check_expiration, only: %i(edit update)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  before_action :get_user, :valid_user, :check_expiration, only: %i(edit update)
```

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  def new; end
+  
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: remove ()

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    else
+      flash.now[:danger] = t ".not_email"
+      render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty?
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: using .blank? instead of empty?

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty? 
+      @user.errors.add(:password, "can't be empty")
+      render :edit
+    elsif @user.update(user_params) 
+      log_in @user
+      @user.update_attribute(:reset_digest, nil)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: remove ()

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      flash[:success] = t ".reset_success."
+      redirect_to @user
+    else
+     render :edit
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: add flash for this case

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    end
+
+    def get_user
+      @user = User.find_by email: params[:email]
+    end
+
+    def valid_user
+      unless (@user && @user.activated? && @user.authenticated?(:reset, params[:id]))
+      redirect_to root_url
+    end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: please, indenting only 2 spaces

File: app/models/user.rb [hunk 7]
```diff
@@ -55,6 +55,20 @@ def send_activation_email
     UserMailer.account_activation(self).deliver_now
   end
 
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_attribute(:reset_digest, User.digest(reset_token))
+    update_attribute(:reset_sent_at, Time.zone.now)
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    update reset_digest: User.digest(reset_token), reset_sent_at: Time.zone.now
```

File: app/models/user.rb [hunk 8]
```diff
@@ -55,6 +55,20 @@ def send_activation_email
...
+  end
+  
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  def password_reset_expired?
+    reset_sent_at < 2.hours.ago
```
Comment (saobv-2223) [inline app/models/user.rb]: please, move 2 to settings.yml file and call here

File: app/views/password_resets/edit.html.erb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<% provide(:title, 'Reset password') %>
```
Comment (saobv-2223) [inline app/views/password_resets/edit.html.erb]: using "

File: app/views/password_resets/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<% provide :title, "Forgot password" %>
+<h1>Forgot password</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:password_reset, url: password_resets_path) do |f| %>
```
Comment (saobv-2223) [inline app/views/password_resets/new.html.erb]: remove ()

File: app/models/user.rb (hunk 7 above)
Comment (chuan97sun) [reply app/models/user.rb]: sorry Mr Bui,the oneline code is not working,i cannot reset my password if use the oneline code

**Comment by chuan97sun** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    else
+     flash.now[:danger] = t ".reset_fail"
+     render :edit
+    end
+  end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:password, :password_confirmation)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: remove ()

**Comment by chuan97sun** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+      redirect_to root_url
+      end
+    end
+
+    def check_expiration
+     if @user.password_reset_expired?
+       flash[:danger] = t ".reset_expired"
+       redirect_to new_password_reset_url
+    end
+  end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  private

  def user_params
    params.require(:user).permit :password, :password_confirmation
  end

  def get_user
    @user = User.find_by email: params[:email]
  end

  def valid_user
    unless @user && @user.activated? && @user.authenticated?(:reset, params[:id])
      redirect_to root_url
    end
  end

  def check_expiration
    if @user.password_reset_expired?
      flash[:danger] = t ".reset_expired"
      redirect_to new_password_reset_url
    end
  end
```

**Comment by chuan97sun** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**